cpp/src/arrow/dataset/discovery.h

```diff
@@ -37,22 +37,54 @@
 namespace arrow {
 namespace dataset {
 
-/// \brief SourceFactory provides a way to inspect/discover a Source's expected
-/// schema before materializing said Source.
+struct InspectOptions {
+  /// See `fragments` property.
+  static constexpr int kInspectAllFragments = -1;
+
+  /// Indicate how many fragments' schema should be inspected. This limits the
+  /// number of fragment accessed to inquire their schemas, thus improving the
+  /// latency of the discovery process when dealing with a high number of fragment
+  /// and/or high latency file systems.
+  ///
+  /// The default value of `1` inspect the schema of the first fragment only,
+  /// in no particular order. If the dataset has a uniform schema for all fragments,
+  /// this default is the optimal value. In order to inspect all fragments, set
+  /// this option to `kInspectAllFragments`. A value of `0` disable inspection
+  /// of fragments.
+  int fragments = 1;
+};
+
+struct FinishOptions {
+  /// Finalize the dataset with this given schema. If the schema is not
+  /// provided, infer the schema via the Inspect, see the `inspect_options`
+  /// property.
+  std::shared_ptr<Schema> schema = NULLPTR;
+
+  /// If the schema is not provided, it will be discovered by passing the
+  /// following options to `DatasetDiscovery::Inspect`.
+  InspectOptions inspect_options{};
+
+  /// Indicate if the given Schema (when inferred), should be validated against
+  /// the fragments' schemas. `inspect_options` will control how many fragments
+  /// are checked.
+  bool validate_fragments = false;
```

Comment:
Naively, for an end user API, I might expect that if you do `validate_fragments`, it will validate all fragments (now, this can still be done on the python side, if that's something we want.

Reply:
Possibly yes.